### ISSUE TITLE
Spike: ABI encoded ICS20 payload

### ISSUE THREAD

**Issue by gjermundgaraba** [cosmos/solidity-ibc-eureka#95]
Context: we currently use JSON encoding the ICS20 payload, which is less than ideal in the EVM because it requires string parsing - and the format of the JSON is very strict (all fields needs to be in the same order every time, no additional fields, etc).

The goal of this task is to implement ABI encoding of the payload structure so that the payload bytes can be directly ABI decoded in the ICS20Transfer contract and see how it affects gas usage. Since a full implementation of this requires ibc-go to also support ABI encoding, and would take a significant amount of time to implement, we should only focus on implementing and testing this on the Solidity side only in this spike. We should still be able to run unit tests with gas reports and get a good indication of the gas cost changes.

Definition of done: a branch with a working ABI encoding of the ICS20 Payload (and all the payload string parsing code deleted) + before/after gas costs of IntegrationTest.t.sol and ICS20TransferTest.t.sol (BenchmarkTest.t.sol is not expected to be working due to the fixtures require a full e2e test to work with ibc-go - which this will not).

**Comment by gjermundgaraba** [comment]
It looks like we would be saving ~20k gas on (250k->230k) send and ~25k gas (286k -> 261k) on receive.

Given that this would need to be implemented on the ibc-go side as well, the next question is if we want to put in the time to do this or not. 

I would recommend we in any case wait for #96, and then we can consider how much lower we need the gas cost and how much effort we are willing to spend on it. 

CC: @womensrights

**Comment by womensrights** [comment]
I think it would be good to scope out fully the possible gains, in other areas and then make a decision based on how much work other gas reductions require. Would be good to understand how much work this would be on the ibc-go side - if its not going to be so big, seems like we should do it. Ultimately any performance gain is going to make Eureka more competitive

**Comment by gjermundgaraba** [comment]
TL;DR:
* 12-15k gas on send
* 25k gas on receive
* ~2 days work to get this fully supported across solidity and ibc-go

I just did a quick test by merging the latest changes which includes the updates to sendPacket and it looks like the cost save on send would be about 12-15k gas and still ~25k gas on receive. This is not entirely unexpected, but gives us a pretty accurate estimate of the benefits.

I did a quick check in ibc-go to gauge the changes needed to support this there, and it looks like a quite small change to get this supported there as well. 

I believe we could have this fully completed in ~2 days.

I think this spike is essentially completed at this point, and if we want to go forward with implementing this, we should create issues in both repositories to do so.

**Comment by gjermundgaraba** [comment]
We have decided to implement this. Will be done in #116
